endpoint.cpp

```diff
@@ -346,35 +349,50 @@ void Endpoint::log_aggregate(unsigned int interval_sec)
     }
 }
 
-int UartEndpoint::open(const char *path, speed_t baudrate)
+UartEndpoint::~UartEndpoint()
+{
+    if (_change_baud_timeout) {
+        Mainloop::get_instance().del_timeout(_change_baud_timeout);
+        _change_baud_timeout = nullptr;
+    }
+}
+
+int UartEndpoint::open(const char *path, unsigned long baudrates[], size_t baud_size)
 {
-    struct termios2 tc;
+    assert(baud_size);
+
+    if (fd > -1) {
+        log_error("UartEndpoint already open");
+        return -EALREADY;
+    }
+
+    _bauds_size = baud_size;
+    _baudrates = baudrates;
 
     fd = ::open(path, O_RDWR|O_NONBLOCK|O_CLOEXEC|O_NOCTTY);
     if (fd < 0) {
         log_error_errno(errno, "Could not open %s (%m)", path);
         return -1;
     }
 
-    bzero(&tc, sizeof(tc));
-
-    if (ioctl(fd, TCGETS2, &tc) == -1) {
+    if (ioctl(fd, TCGETS2, &_tc) == -1) {
```

Comment:
nah... don't store the termios struct. Call TCGETS2 again when you need it (i.e. in the set_speed() method)